pkg/wg/tunnel.go

```diff
@@ -25,6 +32,118 @@ type Tunnel struct {
 	resolv *net.Resolver
 }
 
+func udpPlugboard(c net.Conn, ctx context.Context) (int, error) {
+	laddr := net.UDPAddr{
+		IP:   net.ParseIP("127.0.0.1"),
+		Port: 0,
+	}
+
+	l, err := net.ListenUDP("udp", &laddr)
+	if err != nil {
+		return 0, fmt.Errorf("listen: %w", err)
+	}
+
+	bindAddr := l.LocalAddr()
+	udpBindAddr, ok := bindAddr.(*net.UDPAddr)
+	if !ok {
+		return 0, fmt.Errorf("plugboard: can't recover UDP port")
+	}
+
+	doWrite := func(b []byte) bool {
+		c.SetWriteDeadline(time.Now().Add(5 * time.Second))
+		_, err := c.Write(b)
+		if err != nil {
+			log.Printf("write: %s", err)
+			return false
+		}
+		return true
+	}
+
+	doRead := func(c net.Conn, b []byte) bool {
+		c.SetReadDeadline(time.Now().Add(5 * time.Second))
+		_, err := c.Read(b)
+		if err != nil {
+			log.Printf("read: %s", err)
+			return false
+		}
+		return false
+	}
+
+	var (
+		addr net.Addr
+		lock sync.Mutex
+	)
+
+	readSide := func() {
+		buf := make([]byte, 2000)
+
+		for {
+			if ctx.Err() != nil {
+				return
+			}
+
+			l.SetReadDeadline(time.Now().Add(5 * time.Second))
+			n, a, err := l.ReadFrom(buf)
+			if err != nil {
+				log.Printf("read udp: %s", err)
+				continue
+			}
+
+			lock.Lock()
+			addr = a
+			lock.Unlock()
+
+			var lbuf [4]byte
+			binary.BigEndian.PutUint32(lbuf[:], uint32(n))
+			if !doWrite(lbuf[:]) {
+				continue
+			}
+
+			if !doWrite(buf[:n]) {
+				continue
+			}
+		}
+	}
+
+	writeSide := func() {
+		pbuf := make([]byte, 2000)
+
+		for {
+			if ctx.Err() != nil {
+				return
+			}
+
+			var lbuf [4]byte
+			if !doRead(c, lbuf[:]) {
+				// we're broken here, kill the whole thing
+				continue
```

Comment:
comment is confusing. was this meant to be a return?

Reply:
This is just code I'm goofing around with, if we ever do anything with it it'll look much different.